### ISSUE TITLE
[general need] Synchronisation between mia and capsul engine configuration is suboptimal, in V2

### ISSUE THREAD

**Issue by servoz** [populse/populse_mia#185]
Currently by going to File > Mia preferences it is possible to access the configuration settings for mia but also for capsul engine. In some cases this is completely redundant. 
The synchronisation between the two is currently sub-optimal. We are getting there but I think that a simple user would be lost at the moment.
**It would be necessary to have a perfect synchronisation (instantaneous) in both directions and that one adds or removes a configuration of module.**
Currently it is not all the time easy and faster to do.

In fact we did not invest too much effort in this direction because this duplication should disappear (keep only mia prefrence and a synchronisation invisible to the user would be made at each change, or keep only the display of the capsul engine). 

Another alternative, which goes in the direction of a reflection I had started and which I already mentioned, but which we never settled, would be to have a local or a remote mode of operation (soma-workflow does it, but we don't currently have a defined working mode in mia).
In this case, we could keep both.
The mia preferences would be keep for local use (where it is possible to test the validity of the config, which is currently done for spm and fsl) and capsul engine would be keep for remote use (where it may not be possible to check the config at the time of its definition). But this is beyond the scope of this ticket.

**Comment by servoz** [comment]
After a private discussion with @LStruber, it was decided :
- Until we can do better (perfect synchronisation between Mia's preferences and capsul engine configuration, this ticket), we comment out the CAPSUL part (including the `Edit CAPSUL config` button) in the Pipeline tab of the Mia's preferences.

The Mia's preferences synchronisation with capsul engine configuration works correctly now, but the opposite does not seem to be optimal. So, by commenting on the CAPSUL part, we won't have any more problems on that side. This is only a temporary solution, the best will be to synchronise properly in both directions.

**Comment by LStruber** [comment]
CAPSUL configuration has been commented in commit 0e8c820f601edb7185eab00ae4d564c2da856ab1. I let the ticket open until the synchronisation is completely fixed.

**Comment by denisri** [comment]
Do you have an example / test of bad sync between capsul and mi configs ?
Well, I have added the capsul config because we use it to edit some settings not managed by mia, and now we cannot any longer :/ What's the reason to remove it ? To avoid confusing users ? Can we rather, at least, make it available in admin mode (since this mode exists, let's use it)
Mia is based on Capsul, thus should not reimplement config settings that are already managed by capsul... In a more general way I think all the config should go in capsul config, and have only one config system (whatever GUI we use on top of it, it's another question). Then sync between configs will not be a problem at all anymore.

**Comment by servoz** [comment]
It take this ticket.

**Comment by servoz** [comment]
I found a little time to come back on this ticket.
- playing around fsl config, no system environment variables defined:
   - Nothing defined in Mia preferences:
![Screenshot from 2022-02-10 15-30-15](https://user-images.githubusercontent.com/37933899/153428967-1df99ab7-ffc0-4973-93cf-0b0e265bcde9.png)

  - So nothing is defined in capsul config (good synchronisation):
![Screenshot from 2022-02-10 15-30-32](https://user-images.githubusercontent.com/37933899/153429205-29888a74-08c9-4c7e-94b2-e0aeed744cc8.png)

  - In this case it is not possible to run a FSL process, this is normal.
![Screenshot from 2022-02-10 15-35-18](https://user-images.githubusercontent.com/37933899/153431303-a1a5264d-23bb-4d44-8f19-0c4eccf98361.png)

  - We define the FSL config file in Mia preferences, then we click on the OK button:
 ![toto](https://user-images.githubusercontent.com/37933899/153433074-05d4ae12-8e08-4dfe-8b96-afd9e3d53241.png)

  - If we go back to Mia preferences, then to capsul config (click in the Edit CAPSU config), we can see that the synchronisation has been done with capsul config:
![Screenshot from 2022-02-10 15-58-24](https://user-images.githubusercontent.com/37933899/153434145-06b38c6c-74a9-4032-9add-04d9d0db0aaf.png)

  - Moreover, the FSL processes are now running very well:
![Screenshot from 2022-02-10 16-00-23](https://user-images.githubusercontent.com/37933899/153434860-a898c2cb-e8c3-410c-be6d-646a78f3887d.png)

  - OK, let's say we don't want to do the configuration from Mia preferences but from capsul config. We start from no configuration for FSL:
![Screenshot from 2022-02-10 16-09-49](https://user-images.githubusercontent.com/37933899/153436436-1c81ce49-5f64-4ffa-8ef3-02904fa53baf.png)

  - We click on the Edit CAPSUL config, then we take the FSL tab. Nothing is defined in capsul config (good synchronisation):
![Screenshot from 2022-02-10 16-13-04](https://user-images.githubusercontent.com/37933899/153436932-3d0c3eba-1832-44cf-a8d5-ddf69daeb29c.png)

  - So we fill in the necessary fields (normally it seems to me that capsul automatically deducts the other parameters from the config field, but to avoid any error here the first two are filled in):
![Screenshot from 2022-02-10 16-21-05](https://user-images.githubusercontent.com/37933899/153438503-4a484446-b985-4780-912e-02f3ac929c57.png)

  - Then we click on the OK button:
![Screenshot from 2022-02-10 16-24-16](https://user-images.githubusercontent.com/37933899/153439126-17d4d164-7d11-4014-aaaa-5098cbef7635.png)

- So no synchronisation is done with Mia preference. Maybe after clicking on the OK button? Let's click on the OK button, then we exit the Mia preferences. Let's go back to the MIA preferences:

  - No synchronisation with Mia preferences was done:
![Screenshot from 2022-02-10 16-27-56](https://user-images.githubusercontent.com/37933899/153440226-068f806e-6639-44af-bad3-6e3b82c05f2e.png)

  - In fact even in capsul config the FSL config is not here:
![Screenshot from 2022-02-10 16-28-16](https://user-images.githubusercontent.com/37933899/153440579-8e95e34e-ee33-4b06-9e58-5b523039afbc.png)


Moreover it seems that the configuration is not kept when the GUI of capsul config is closed by clicking on the OK button because if, without leaving the Mia preferences, we click again on Edit CAPSUL config button, the FSL configuration is already lost. I will investigate this. However, I am away for a fortnight from tomorrow.

**Comment by servoz** [comment]
It's fixed for FSL.
If I remember well (I haven't retested yet) there was also a problem with matlab /spm.
I'll have a look later, in two weeks.

**Comment by servoz** [comment]
I start to see for Matlab/SPM.
Currently, 

-1 if nothing is defined in Mia preferences,
![rien](https://user-images.githubusercontent.com/37933899/156001809-7d142a4e-7d9f-4d3c-b8dc-79b4be5c447e.png)
we observe in the GUI of capsul config:
![matlab_rien](https://user-images.githubusercontent.com/37933899/156000292-ca7e6219-10ab-40b6-aae8-fe2d94eeaadc.png)
![spm_rien](https://user-images.githubusercontent.com/37933899/156000306-db47eb24-f90d-475e-881a-710bd7b9fc3f.png)
This looks good!

-2 Now if we define Matlab/SPM in mia preferences:
![mia_pref_matlab](https://user-images.githubusercontent.com/37933899/156001393-ab234bd1-d1a0-4d75-ac4d-627beeec15b9.png)
we observe in the GUI of capsul config:
![matlab](https://user-images.githubusercontent.com/37933899/156002550-26bde71c-0386-4abe-9cfb-426926def9a3.png)
![spm](https://user-images.githubusercontent.com/37933899/156002567-17569260-fde2-4673-944d-c0dcc4f63f17.png)
This looks good!

-3 Now if we define MCR/SPM standalone in mia preferences:
![mia_pref_mcr](https://user-images.githubusercontent.com/37933899/156003148-5dcbf678-6943-4a1e-ab9a-57da586b20ae.png)
we observe in the GUI of capsul config:
![matlabstandalone](https://user-images.githubusercontent.com/37933899/156003210-a74ad2c9-aa5f-4b48-b510-c851c5ab93ba.png)
![spmstandalone](https://user-images.githubusercontent.com/37933899/156003222-8551ab38-4928-4125-921f-ca3a98edd058.png)
This looks good for spm and wrong for matlab. Am I wrong? 
How should the value of the matlab executable field be in this case? /home/econdami/MATLAB/MATLAB_Runtime/v95 ?

**Comment by servoz** [comment]
In fact in capsul/study_config/config_modules/matlab_config.py, matlab_exec is defined as a File trait. So it is not possible actually to give here the path to the MCR (a Dir). So it is not possible to define the MCR path in the GUI of capsul config and it happens elsewhere in populse_mia, under the hood:

-> In populse_mia/software_properties.py (line 863):
```
       if capsul_config.get('use_spm', False):
            spm_dir = capsul_config.get('spm_directory')
            spm_standalone = capsul_config.get('spm_standalone')
            #TODO: I thing the following is wrong
            if spm_standalone:
                mcr = os.path.join(spm_dir, 'mcr', 'v713')
                if os.path.isdir(mcr) and os.path.isdir(spm_dir):
                    self.set_spm_standalone_path(spm_dir)
                    self.set_matlab_standalone_path(mcr)
                    self.set_use_spm_standalone(True)
                    self.set_use_matlab_standalone(True)
                    self.set_use_matlab(False)
            else:
                self.set_use_spm_standalone(False)
                if self.get_use_matlab():
                    self.set_spm_path(spm_dir)
                    self.set_use_spm(True)
                else:
                    self.set_use_spm(False)
```

As it is, it is not possible to synchronise capsul config with mia preferences, and worse, we have to use only mia pref to define the MCR path.

This is where I reach the limit of what I can do to close this ticket. I think the capsul config GUI should be able to accept a file or a directory for "executable" fields in the matlab tab. In this case, I'm not sure if "executable" is really suitable in the GUI ("path" would be better?).
Anyway, I need you to tell me what you have imagined on the subject. For now the whole thing works by chance for a bad reason (the hard-coded line: mcr = os.path.join(spm_dir, 'mcr', 'v713', see above)). If we want to put the capsul config GUI back in use we have to fix this problem. Then the synchronisation should work, with some modifications of the populse_mia codes.
I work in the with_capsul_config branch, where the Edit capsul config GUI is available.

Translated with www.DeepL.com/Translator (free version)

**Comment by servoz** [comment]
Of course! This is exactly what we did in mia pref. There is a matlab_exec and a matlab_standalone.
I respect your point of view, but, concerning matlab, whether it is an executable or not it is a file for an executable or a path in case of an MCR. It is more or less the same for SPM. In the case of standalone it is a file and in the case of "normal" SPM it is a directory ... 
What would be nice is to be able to define in the GUI either the matlab_exec or the matlab_standalone.

**Comment by servoz** [comment]
For SPM we use in both cases the "directory" field, so it's ok for the file (.sh) or the directory (in fact we add the .sh, under the hood, to the directory. But for matlab as you want to keep the notion of "executable" it's less obvious (for me we need the directory for the matlab executable and like for spm we could add, under the hood). With this, a "directory" field is enough instead of the "executable" field. But if you have another quick solution, I am interested.
For the moment I'm stuck until we know how to do it in the capsul side, to define the bin/matlab in the case of matlab or the path to the MCR in the case of the Compiler Runtime).
Also, I believe that it is necessary to make simple for lambda users. For us the notions of executable or MCR are important. But for a lambda user I am not sure that it is very important!

**Comment by denisri** [comment]
Why use the same field for matlab executable and the MCR ? If both can be used to run SPM, in a more general way it is not the same at all: you can run matlab alone, you can't run the MCR (I guess); one is a file, the other is a directory. Thus it seems simpler and clearer for me to use two config fields. However if you really want to, we can use the traits.Either type to allow a file or a directory, but I'm not convinced it its the best solution.
For me I would add a field matlab_mcr in capsul config, and if it were just for me I would completely remove it from MIA's config, to avoid redundancies, even if we keep MIA's GUI to edit it: its specialized GUI is obviously better than the generic GUI in Capsul, but it could allow to edit the same single config.
But as you like... Tell me if you want me to add the missing field in Capsul...

**Comment by servoz** [comment]
I think that MCR can be used without spm standalone (for example if we want to use compiled matlab codes, I have to look into it as soon as possible because I have matlab codes that I want to use without access to the source codes). But this is not the question.
if it's possible to define both, it's ok for me. it's just that i don't see yet exactly how it will work to find it (for spm it's easy, you check standalone or not, so you know exactly what it means) ...

**Comment by servoz** [comment]
I don't think it's a good idea to remove matlab_mcr from mia pref. The idea is to be perfectly synchronised, that is to say to be able to define the same things in mia pref and capsul config .... and for the moment to define the MCR we are obliged to go through mia pref, capsul config does not allow it.

**Comment by denisri** [comment]
> I don't think it's a good idea to remove matlab_mcr from mia pref

Why not ? Why is it a good thing to duplicate config options ?

> capsul config does not allow it.

that's exactly why I ask if I should add it in capsul config.

The reason I insist is, in addition to not liking duplicating things, that, for me MIA GUI should be a graphical interface to run Capsul pipelines, which we should normally be able to run also outside of the GUI, typically to run them on a computing resource, from a script or something else. Capsul may allow extensions like the features of MIAProcesses, databasing and provenance tracking etc. There are many things that were developed inside MIA's GUI (see the pipeline_management_tab.py which is both the GUI and in a large part also processing infrastructure. I would like to separate them, probably not in the short term, but we want, at the end, to be able to run any processing from the commandline `python -m capsul` (or what it becomes in the future). If we keep things separated like today, with a config for capsul, and another for MIA, a separate processing/run engine/loop, etc, we will not reach this goal, which is for me a deal-breaker for the whole project.

**Comment by servoz** [comment]
To tell the truth, if the GUI of capsul config works as well as mia pref, I have no problem to delete mia pref ... The actual problem goes beyond the synchronisation issue ... The problem is that capsul config doesn't work for the MCR. It's mandatory to use mia pref ... That's why we have unplugged the GUI of capsul config in mia pref in master actually ...

**Comment by denisri** [comment]
I have added the matlab `mcr_directory` field in capsul config.

**Comment by servoz** [comment]
Thanks @denisri ! I will see ASAP how to deal with ...

**Comment by servoz** [comment]
it is still not synchronised and it is still necessary to go through the mia prefs to update the GUI capsul config and thus allow to launch a calculation with SPM standalone.
But now I "only" have to manage the synchronisation in both directions, in populse_mia.
I maintain that having both definitions possible at the same time (one can never prevent a user from putting both!) is a source of error! which one to take?
I will try to manage this in populse_mia!

**Comment by denisri** [comment]
Hmmm... It is synchronized except that... 
* there is a bug somewhere in the capsul config GUI:
if I edit Capsul config, set matlab mcr_directory, and... wait a couple of seconds before validating, then the new value gets in the MIA settings. But if I click "OK" quickly, then the value is not updated. So I guess it lacks a forced validation of capsul GUI. I have to look at it.
* the other way, if I edit in MIA the value of matlab standalione path, then click "edit capsul config" *before validating MIA config*, then the value visible in capsul is the former one, because Capsul config is built from the state of MIA config, and not from the GUI, and MIA config is not changed until it is validated. We could hack this if needed, but as you said we should rather keep only one way to edit a given option, so in the long run we should rather hide the matlab/spm part of capsul config GUI.

**Comment by denisri** [comment]
The 1st way is fixed normally. The other way probably requires more work since it needs to take values in MIA config GUI, not in the config structure itself.

**Comment by servoz** [comment]
it is not synchronised.
Maybe we are not talking about the same thing? 
To me, synchronised means that we can edit the configuration in both directions. From mia pref to capsul config and from capsul config to mia pref. This is not the case for matlab/spm. 
I will look into this as soon as possible (I have already done it for FSL, vide supra)

**Comment by servoz** [comment]
I have a very hard time with this ticket. Most of the problems are mainly due to issue in the capsul config gui. I have started to plan some patches on a capsul branch. However, there are two things that if you had a solution, would save me from doing introspection in soma and capsul (which is interesting for the knowledge of the machinery, but is very time consuming !!!).

1- It is not possible to delete a parameter value. Let's say we want to switch from SPM to SPM standalone. We have as a starting point for matlab parameters:
![Screenshot from 2022-03-04 11-47-15](https://user-images.githubusercontent.com/37933899/156749817-67456e72-0700-4dee-a153-81200f327454.png)
We define the mcr_directory:
![Screenshot from 2022-03-04 12-59-36](https://user-images.githubusercontent.com/37933899/156759767-fd22d6d8-6231-458a-8dbd-b4024acd2831.png)

To be clean we want to remove the value of executable (there may be other reasons to remove the value of a parameter). 
![Screenshot from 2022-03-04 11-54-03](https://user-images.githubusercontent.com/37933899/156750556-dffefaa1-2fbb-451f-9378-6af1e823b902.png)
This is not possible. The old value automatically reverts (in this case we would like to revert to `<undefined>`)

2- It is not posssible to define mcr/standalone parameters for matlab/spm:
Let say we start with a empty config:

*matlab/spm case:*
![Screenshot from 2022-03-04 12-08-48](https://user-images.githubusercontent.com/37933899/156753000-8156e73d-5394-4b92-9faf-04b8d38059df.png)
We have the strange idea to change config from casul config gui :-):
![Screenshot from 2022-03-04 12-12-59](https://user-images.githubusercontent.com/37933899/156753523-f83a9b84-2d8e-491c-88bf-ece8bf3af19b.png)
![Screenshot from 2022-03-04 12-13-35](https://user-images.githubusercontent.com/37933899/156753536-9b396785-670a-4803-b53a-89d9b5d4a635.png)
We click on the OK button.
All is fine:
![Screenshot from 2022-03-04 12-19-49](https://user-images.githubusercontent.com/37933899/156754313-3291558b-154a-4ffb-9c54-6571cdfe678b.png)

*mcr/standalone case*
We start again from a empty config.
In  the capsul config gui:
![Screenshot from 2022-03-04 12-59-36](https://user-images.githubusercontent.com/37933899/156759946-a6395d1c-8292-4459-800f-6fec477885f1.png)
![Screenshot from 2022-03-04 12-24-05](https://user-images.githubusercontent.com/37933899/156755012-d8aaf558-8ee3-4f39-a9aa-0bb378a0ad96.png)
We click on the OK button, and we stay with a empty config in mia config.

In fact the issue here does not come from a synchronisation problem between capsul config and mia config, but rather from the capsul/soma side (well I think).
Indeed for matlab part we see in capsul/engine/module/matlab.py, introspection:
![Screenshot from 2022-03-04 12-37-52](https://user-images.githubusercontent.com/37933899/156756858-0b85ed59-91e4-4cc8-9365-910ee6ca49a3.png)
conf is not None and executable have the good value in controller object.
BUT for spm part, we see in capsul/engine/module/spm.py, introspection:
![Screenshot from 2022-03-04 12-46-10](https://user-images.githubusercontent.com/37933899/156757882-c0916375-77fc-445d-a745-54012e094d8a.png)
directory and standalone attributes are ok in the controller objetc but the conf is None ... 
So we lost the values ... And the final result is:
![Screenshot from 2022-03-04 12-54-28](https://user-images.githubusercontent.com/37933899/156759058-3cf2f0d7-9157-4b1a-910e-06bdc13777a7.png)

It's a pretty painful ticket because before working on the synchronisation you have to debug the capsul/soma side !!!

**Comment by denisri** [comment]
1. It's not very intuitive I admit, but to reset the value you have to type in the field the string `<undefined>`: emptying the edit line will just cancel edition and get back to the older value. The Controller GUI will be redesigned anyway.
2. You have removed yourself (commented out, to be exact) the line which performed the sync:
    https://github.com/populse/populse_mia/commit/381dd59f03d5d7cbc9b347bed20b33b85d792913
    by commenting out the line:
```
         self.set_matlab_standalone_path(mcr_dir)
```
    you forbid setting into MIA the edited Capsul config option for matlab standalone path.
    When I uncomment this line, it works (again) for me.

**Comment by servoz** [comment]
for -1 I had seen that <undefined> allowed a reset, but I was hoping for something more intuitive. Ok let's keep it that way, especially if the gui controller needs to be redesigned!
for -2 yes I know I commented on this, but it seemed to me that the problem came from well before (as described in the previous post) the set_capsul_config method of the populse_mia.software_properties module...
Ok I'll look into it right away

**Comment by servoz** [comment]
Ok ... I'm sorry I was on something else and had already forgotten about the thread...

In fact for 2- the problem does not come from matlab but from spm (that's why I haven't uncommented `self.set_matlab_standalone_path(mcr_dir)` because before working on the synchronisation it is necessary that the data edited in the gui of capsul config are well available ....

Look  in my last post above I never mentioned matlab but spm ... 

To be very clear this is what we have when we pass trough the set_capsul_config method:

*engine_config have not the good value for spm:*
![Screenshot from 2022-03-04 15-56-35](https://user-images.githubusercontent.com/37933899/156786086-3c0f1b84-1185-4f77-90f6-1d23ce05d377.png)

*Same thing for the capsul_config object:*
![Screenshot from 2022-03-04 15-58-13](https://user-images.githubusercontent.com/37933899/156786385-d6a5234d-dbe9-4f22-ba14-0dc971234905.png)

**Comment by servoz** [comment]
By the way, if I uncomment s`elf.set_matlab_standalone_path(mcr_dir)` it doesn't change anything on the result. This is not surprising because the edition of the spm parameters is not preserved...

**Comment by servoz** [comment]
The good question is why we lost the edited spm standalone parameters ?

**Comment by servoz** [comment]
Ok, I've already spent a lot of time on this ticket. I can't understand why the spm standalone values are not kept in the capsul config engine after a `Edit CASPUL config` in mia preferences. Fortunately, the equivalent on the mia preferences side works correctly. Since the `Edit CAPSUL config` button is no longer connected in the master branch of populse_mia, I think the best thing to do is to close this ticket !

**Comment by denisri** [comment]
I have done modifs to remove the use of the old study_config in this part, and to fix a number of issues in capsul/mia config sync. I continue to believe that Mia should take its config from capsul directly rather than duplicate everything in a different shape, because it's one of the (multiple) factors which make sync complex and "fragile", but I haven't worked on that side. In a general way, all of Mia config should be Capsul config modules, no more no less.

Anyway you have to understand that we have added the Capsul config GUI for a reasons, it's not a caprice: some config options do not exist in Mia and can be set only from Capsul config Gui, so we _need_ it. Putting it in a dead branch and closing the ticket is not an option. Plus, there is no modularity in Mia config. We need modularity.

**Comment by servoz** [comment]
Maybe you should be careful because if I didn't know you I would think that your last post is very arrogant...

Anyway, you must also understand that it is not possible to put things in Mia that do not work, or that will work one day ... Even if this thing, which does not work, is extremely important for you...

To say exactly and calmly what I think, I doubt we can go on like this for long. I can't spend my time trying to fix what's wrong with Mia after the changes, very important for you, that you made. This without any consultation, since you never have the time to have a meeting for a long time. It is true that we in Grenoble have plenty of time, when we are not skiing or playing tennis.

Ah just a last things ... since your last change, mia crash when we click in the `Edit CAPSUL config` button of the mia pref ... in addition to not working, it makes Mia crash. It's perfect!

**Comment by denisri** [comment]
It does not crash for me. Maybe things have not be updated for all projects ?
Anyway Mia is in development, we have different setups and environments, there are bugs (and not only on my side, but yes, like everyone, I sometimes make mistakes), things sometimes have to be fixed, and we do it. Now saying "this does not work, I remove it", I'm not very found of it. We clearly don't have the same objectives, and it's difficult to keep a common direction. Maybe too difficult. I also have heavy doubts on the project.

**Comment by servoz** [comment]
>> I also have heavy doubts on the project.

it is not impossible that this is the last thing we agree on.

**Comment by sapetnioc** [comment]
Don't you think it is time to have a meeting to talk about the future
direction of the project ? We worked a lot together with a good atmosphere
and mutual understanding. If this is no longer the case, we must find a
solution so that everyone can follow their objectives with serenity. Let's
not let an IT project completely ruin our human relations.

**Comment by denisri** [comment]
We can talk, sure, for me I'm not angry.

**Comment by servoz** [comment]
Planning a meeting is the best we can do at the moment, I think.

**Comment by servoz** [comment]
If we have defined a configuration for spm and spm standalone:
In software.properties.get_capsul_config() we pass through (l311 with_capsul_config branch):
```
       if use_spm:
            m = eeconf.setdefault('capsul.engine.module.spm',
                              {}).setdefault(spmconf['config_id'], {})
```

but  spmconf is not defined and raise exception:

```
Traceback (most recent call last):
  File "/data/Git_Projects/populse_mia/python/populse_mia/main.py", line 1313, in <module>
    main()
  File "/data/Git_Projects/populse_mia/python/populse_mia/main.py", line 812, in main
    config.get_capsul_engine()
  File "/data/Git_Projects/populse_mia/python/populse_mia/software_properties.py", line 389, in get_capsul_engine
    capsul_config = config.get_capsul_config()
  File "/data/Git_Projects/populse_mia/python/populse_mia/software_properties.py", line 313, in get_capsul_config
    {}).setdefault(spmconf['config_id'], {})
NameError: name 'spmconf' is not defined




```

**Comment by servoz** [comment]
I guess it is a typo and we can change by?:
 ```
if use_spm:
            m = eeconf.setdefault('capsul.engine.module.spm',
                              {}).setdefault('spm', {})

```

**Comment by servoz** [comment]
I had rather gone on the #191  ticket (which is long and tedious because it is about code cleaning ...). I'll come back to this ticket (#185) very quickly to try to understand what's wrong (tomorrow is already quite busy, but it's a priority for me). We'll keep in touch !

**Comment by denisri** [comment]
Well I don't have the same, and I don't know what's different.
All projects updated and on the same branches as you show.
When I erase the config files (`properties/config.yml` especially) and start populse_mia, **I don't have the message**:
```
An issue is detected in the Mia's configuration:
dictionary update sequence element #0 has length 1; 2 is required
Please check the settings in File > Mia Preferences > Pipeline ...
```
When I erase `process_config.yml` I have a very long log because all processes are re-scanned, and it should not matter here. When I only erase `config.yaml`, the output is:
```
$ python3 -m populse_mia

Checking the installed version for nipype, mia_processes and capsul ...

** mia is already using the current installed version of nipype, mia_processes and capsul for this station (1.7.0, 2.0.1-dev+dd90bd8 and 2.3.0, respectively)

QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-dr144257'
```
nothing more at startup.
I open MIA preferences, go to admin mode, and click "edit config", I only have:
```
admin_hash: 60cfd1916033576b0f2368603fe612fb78b8c20e4f5ad9cf39c9cf7e912dd282
mia_user_path: /casa/host/src/populse/populse_mia/master
opened_projects:
- /tmp/temp_mia_project1q84smsi
source_image_dir: /casa/host/src/populse/populse_mia/master/python/populse_mia/sources_images
```
so at this stage I don't have any Capsul config at all.
You seem to have a remaining Capsul config. Where can it come from ? I don't know if this is the cause of the problem.
* When I restart populse_mia and edit again the config (without having changed anything else, and I have to re-type the admin password since I have not saved the config), I now have:
```
admin_hash: 60cfd1916033576b0f2368603fe612fb78b8c20e4f5ad9cf39c9cf7e912dd282
capsul_config:
  engine:
    global:
      capsul.engine.module.attributes:
        attributes:
          attributes_schema_paths:
          - capsul.attributes.completion_engine_factory
          - populse_mia.user_interface.pipeline_manager.process_mia
          config_environment: global
          config_id: attributes
          process_completion: mia_completion
  engine_modules:
  - nipype
  - fsl
  - freesurfer
  - matlab
  - spm
  - fom
  - python
  - afni
  - ants
mia_user_path: /casa/host/src/populse/populse_mia/master
opened_projects:
- /tmp/temp_mia_projectn3d5cxif
source_image_dir: /casa/host/src/populse/populse_mia/master/python/populse_mia/sources_images
```
* Then I go to the pipeline tab (of the preferences), everything is empty/unchecked (no need to upload a snapshot).
* I check "use matlab standalone", set the path for it: `/host/usr/local/spm12-standalone/mcr/v713` for me, check "use SPM standalone" and enter the path: `/host/usr/local/spm12-standalone`.
Click "OK" -> I get a warning popup for `MRIFileManager.jar` but otherwise OK. Nothing more in `stdout/stderr`.
* go again in File > Mia preferences > Tools tab, I click in Edit config:
```
admin_hash: 60cfd1916033576b0f2368603fe612fb78b8c20e4f5ad9cf39c9cf7e912dd282
auto_save: false
background_color: ''
capsul_config:
  engine:
    global:
      capsul.engine.module.attributes:
        attributes:
          attributes_schema_paths:
          - capsul.attributes.completion_engine_factory
          - populse_mia.user_interface.pipeline_manager.process_mia
          attributes_schemas: {}
          config_environment: global
          config_id: attributes
          process_completion: mia_completion
      capsul.engine.module.axon:
        axon:
          config_environment: global
          config_id: axon
          shared_directory: /casa/host/build/share/brainvisa-share-5.1
          user_level: 0
      capsul.engine.module.fom:
        fom:
          auto_fom: true
          config_environment: global
          config_id: fom
          fom_path: []
          input_directory: /tmp/temp_mia_projectn3d5cxif/data/raw_data
          output_directory: /tmp/temp_mia_projectn3d5cxif/data/derived_data
      capsul.engine.module.freesurfer:
        freesurfer:
          config_environment: global
          config_id: freesurfer
      capsul.engine.module.matlab:
        matlab:
          config_environment: global
          config_id: matlab
          mcr_directory: /host/usr/local/spm12-standalone/mcr/v713
      capsul.engine.module.nipype:
        nipype:
          config_environment: global
          config_id: nipype
      capsul.engine.module.spm:
        spm:
          config_environment: global
          config_id: spm
          directory: /host/usr/local/spm12-standalone
          standalone: true
      capsul_engine:
        uses:
          capsul.engine.module.afni: ALL
          capsul.engine.module.ants: ALL
          capsul.engine.module.attributes: ALL
          capsul.engine.module.axon: ALL
          capsul.engine.module.fom: ALL
          capsul.engine.module.freesurfer: ALL
          capsul.engine.module.fsl: ALL
          capsul.engine.module.matlab: ALL
          capsul.engine.module.nipype: ALL
          capsul.engine.module.python: ALL
          capsul.engine.module.spm: ALL
  engine_modules:
  - nipype
  - fsl
  - freesurfer
  - matlab
  - spm
  - fom
  - python
  - afni
  - ants
clinical_mode: false
control_V1: false
mainwindow_maximized: true
mainwindow_size:
- 0
- 0
matlab_standalone: /host/usr/local/spm12-standalone/mcr/v713
max_projects: 5
max_thumbnails: 5
mia_user_path: /casa/host/src/populse/populse_mia/master
mri_conv_path: ''
opened_projects:
- /tmp/temp_mia_projectn3d5cxif
projects_save_path: /casa/host/src/populse/populse_mia/master/projects
radio_view: true
source_image_dir: /casa/host/src/populse/populse_mia/master/python/populse_mia/sources_images
spm_standalone: /host/usr/local/spm12-standalone
text_color: ''
use_afni: false
use_ants: false
use_fsl: false
use_matlab: false
use_matlab_standalone: true
use_spm: false
use_spm_standalone: true
user_mode: false
```
Both MIA and Capsul configs have been updated.
* I go to the Pipeline tab:
* I click on Edit Capsul config:
the GUI pops up, and displays the correct values.
* If I edit them to change matlab mcr_directory and spm_standalone directory, for instance, and validate, I see the new values in MIA.

So there is definitely something different beween your setup and mine....

The error message "An issue is detected in the Mia's configuration" is issued when setting Capsul config values from the dictionary of the previously existiing config, so there is probably something wrong left in your config, but I don't understand what nor where it comes from since you have erased your config before starting...

**Comment by servoz** [comment]
Ok, thanks for your comments! !!! This is a crazy story!!!!
 I'll take a look at it and try to figure out what's wrong ASAP (maybe tomorrow, but I already have a busy day, Thursday at the latest!)

**Comment by servoz** [comment]
Um, it seems , in your case, that mia doesn't start in developer mode, I think because populse_mia is in /casa/host/src/populse/populse_mia/master.
If this is the case, I will have to adapt main.py to handle this case which I had not thought of.
To be sure that it is indeed that (I do not see what else it could be, a priori, since we delete the config file and have populse up to date and on the same branches!), can an you please do the following test:
Clone capsul, populse_mia, etc (all populse) in a directory for example ~/tototest. Then launch mia from the main.py of this directory (python3 main.py).
On my side I will test by putting populse_mia in casa/host/src/populse/populse_mia/master and start with python3 -m ... I think and I hope that the problem comes from this ....

Translated with www.DeepL.com/Translator (free version)

**Comment by denisri** [comment]
OK, I just pushed a fix. The problem was actually in `main.py`: it detects a few hard-coded paths for the different projects, sets up `sys.path`, and stores a CapsulEngine python config for these paths. The config dict was missing a dictionary level (actually I have added a level to store several configs as it is allowed in Capsul for instance to store several SPM configs, which is actually used). And thus it did not have the expected format when imported in CapsulEngine.

Is this "developer mode" still useful ?

**Comment by denisri** [comment]
And now it fails using the non-dev mode :/

**Comment by servoz** [comment]
I'm also looking at the main module (but not for long, I have numerous meetings all afternoon, I'll start again tomorrow morning) and I think it's necessary to clean up the DEV and USER parts of this module which I think are no more really adapted with capsul engine (even more so if we remove study_config, as we are doing now).

In a nutshell (if I remember correctly, because it's been a while since we did this), the DEV mode is rather dedicated to run from a directory where populse packages are cloned/installed. Basically without going into details it allows to add to the sys.path the packages of the directory and without using PYTHONPATH. I know you have another way of working with casa_distro. The best would be to keep everyone's habits by making main handle it. That's why I'm sure we'll find a solution to adapt the main module to work in all cases (basically: in pure user mode - everything is installed in the site-packages and run with python -m populse_mia - , in pure dev mode - everything is installed in a personal directory and run with python main.py - in dev mode casa_distro - everything is installed in /casa/host/src/populse/ and run with python -m populse_mia.

Concerning the size of the populse_mia package, I totally agree it is too heavy while the codes are indeed very light. It's also an old problem to which we never found a simple answer until now. There are several options but we have never gone to the end of the thought process and made a decision in this sense. I agree that it is time to make a decision. For me it should be that the user has the choice to get only the codes or the codes + the data if he wants to use them. It is also necessary that the data are present for the UT. I had started to look at what git-lfs can do for that, I have a vague memory that it was possible but I confess I did not take the time to go to the end of things in this sense. Maybe there is another solution ?

**Comment by servoz** [comment]
It's true that I like working in a development folder, but maybe it's just a habit... Let's give it some time (let's say until tomorrow night, because I'll have to stop for today and start again tomorrow) and if it seems too difficult, I'll put populse_mia and mia_processes in casa_distro and in that case there won't really be any need for dev mode ... (in this case I might not be the only one to change my habits, maybe Lucas ? others on Grenoble ? but it is not impossible to do ...). But in this case we won't be able to develop outside casa_distro ... Well, we will find a solution in this case.

**Comment by denisri** [comment]
> And now it fails using the non-dev mode :/

Ah OK I had not erased the config file after switching branches, thus an incompatible config was loaded. I will try to make it clearer and non-fatal.

**Comment by denisri** [comment]
> It's true that I like working in a development folder, but maybe it's just a habit... Let's give it some time (let's say until tomorrow night, because I'll have to stop for today and start again tomorrow) and if it seems too difficult, I'll put populse_mia and mia_processes in casa_distro and in that case there won't really be any need for dev mode ... (in this case I might not be the only one to change my habits, maybe Lucas ? others on Grenoble ? but it is not impossible to do ...). But in this case we won't be able to develop outside casa_distro ... Well, we will find a solution in this case.

If it's just a matter of detecting paths, no need to change. But maybe we don't need to store them permanently in the config file, since they are detected again the next time we run `main.py` anyway.

**Comment by denisri** [comment]
I'll open another ticket for data, since it's not related.

**Comment by denisri** [comment]
> I'll open another ticket for data, since it's not related.

see #260.

**Comment by servoz** [comment]
Currently we observe in stdout (dev mode; populse in a personal folder, start from python main.py):
```
An issue is detected in the Mia's configuration:
import_configs() got an unexpected keyword argument 'cont_on_error'
Please check the settings in File > Mia Preferences > Pipeline ...
```
[I propose this very little change.](https://github.com/populse/capsul/pull/218)

**Comment by denisri** [comment]
Yes sorry I did the same change on my side but apparently forgot to commit/push it (because I have several works on the way at the same time). Thanks !

**Comment by servoz** [comment]
no problem, I didn't do much!

**Comment by servoz** [comment]
Currently, if no configuration exists for spm, the capsul engine GUI defaults to the standalone checkbox and uses version 12.

I think, by default the standalone box should not be checked and the version should be undefined if the module configuration is not known.

If it doesn't contradict something I wouldn't see, [I propose this little PR ](https://github.com/populse/capsul/pull/219)along this line (version = undefined and standalone =False)

**Comment by denisri** [comment]
Not having the standalone version by default assumes a much more expensive dependency (matlab) by default, but I don't really mind: the user just has to configure it. For the version OK for me, I just wonder if it will leave an additional config with an "no-version" ID, but it's probably not very important.

**Comment by servoz** [comment]
The capsul config settings validation  is protected by a `try` for [matlab parameters](https://github.com/populse/capsul/blob/master/capsul/engine/module/matlab.py#L48), which is [not the case for spm](https://github.com/populse/capsul/blob/master/capsul/engine/module/spm.py#L168).

For matlab it looks more like a "try of development" to allow not to crash but to have access to the traceback in case of exception raising.

The resolution of this ticket exceeds the only synchronisation between mia preference and capsul config (we want something that works in all cases from mia pref to capsul config and from capsul config to mia pref, that does not crash and that handles all possible cases).

The GUI of capsul config is for the moment very rustic. For example Mia crashes if a wrong parameter is entered (and only for some parameters). I don't go into details, but I will propose a small PR for that very soon. One solution is to try to protect certain parameter changes. So I'm thinking of deleting the first `try` I mentioned at the beginning of this post (it will be handled differently, and in a more silent way).

Does it make a problem for capsul (side effects I don't think?)

**Comment by denisri** [comment]
if I remember, the `try`.. `except` was because the `validate_config()` function is called within a GUI callback, and an exception raised here and not catched ends up with the whole application aborting. I don't remember if the try could be put somewhere else in a higher level function, still inside the GUI callback, but it certainly has to exist somewhere. The traceback print is mainly for debug purpose, yes, it could probably be replaced with something "nicer", but anyway the user should be informed that something has failed in the values validation, and what/why.
If it can be done at higher level, we could perhaps have a single try...except. If not (I'm sorry I don't have time to investigate this today), then every module (matlab, spm and others) would have to do it, because if not, any invalid value entered by the user would cause a complete application abort, and of course it is not what we want in Mia.

**Comment by servoz** [comment]
Sure, I totally agree!

That's why I propose to remove this `try` (which exists only for the matlab module, currently) and protect the validation for all modules at a higher level; in capsul.qt_gui.widgets.settings_editors.SettingsEditor.accept().
I made this change, and others in the capsul [GUI_editCapsul branch](https://github.com/populse/capsul/commits/GUI_editCapsul).
I will make a PR from this branch very soon.

I'm waiting until I finish working on this ticket to make it, so I don't have to make multiple PRs.
ANTS, FSL, AFNI, MATLAB and SPM are now synchronised from mia pref to capsul config. I still have some work to do from capsul config to mia pref.

It will be necessary of course to add in Mia pref other modules config, as FreeSurfer, etc. I think that for nowe, given the number of priority things to do, the best is to add the modules when needed, as @LStruber did with ANTS and AFNI, when he needed them to make the MRIQC pipeline.

**Comment by servoz** [comment]
I'm coming to the part of the work on synchronisation from capsul config to mia pref.

We recently added a new parameter in capsul config to differentiate between the executable (executable parameter) and the MCR directory (mcr_directory parameter).

This is a very good thing to be as close to reality as possible.

However, [as I said](https://github.com/populse/populse_mia/issues/185#issuecomment-1055427418), for the sake of simplicity (especially for the average user), it is sometimes better not to be too precise. I was also pointing out that for SPM we already have a directory parameter which will ultimately be either a directory in the case of spm, or an executable, after adding the .sh file. So for spm things are simple, the user only fills in the directory field in capsul config and implicitly chooses between spm or spm standalone.

On the other hand, as we have two parameters for matlab, things get more complicated to synchronise with mia pref, and even to know what to choose if the user fills in both parameters (we will use one, matlab, or the other MCR?).

If the user sets the SPM config at the same time, we can deduce which parameter to use. On the other hand, if the user does not define SPM? This is the case where we want to use matlab without spm or MCR without spm standalone .....

We are in a quandary ....

I will only retain matlab to MCR if both are defined without concomitant definition of spm, but there we lack precision (no, no, I don't want to reopen the discussion on the usefulness of having two parameters for matlab, executable and mcr_directory, although ... maybe.. :-))).

**Comment by servoz** [comment]
After many headaches, and a lot of work, I think we're coming to the end of the processing of this ticket.

It remains to be tested if it really works properly.
It took me quite a while to figure out what was wrong, often the result was not easily understandable because of the machinery (which in my opinion is not completely finished) to set up the capsul config. I won't go into details, but this sometimes led to something that looked more like a patch... Maybe there is no need to do more than a patch as Capsul config will certainly be reworked very soon.

Anyway, I'm afraid this could still lead to unexpected results ...

The goal was to have a **perfect synchronisation between mia pref and capsul config**. That is:
If we open File > Mia preferences; Pipeline tab:
  - If we hit Edit CAPSUL config button, we must find exactly the same configuration as in Mia pref.
  - If we click on OK, we should find in Mia pref exactly the same configuration as in capsul config.

In most cases this should work now. There is one more case that could be a problem: 
In a nutshell, Capsul config allows multiple settings for one module. e.g. spm, spm12, etc ... But the Capsul config GUI only shows the first item in the dictionary ({engine: {global: {capsul.engine.module.spm: {spm: ...}}}} in the previous example ... This can lead to hidden configurations and therefore to unexpected results ... I tried to do the best I could with the existing ...

So, please, test a bit the `with_capsul_config` branch and play with the `Mia pref` to the `Capsul config` and see if we have a good synchronisation and report here if you observe any problems still existing.

I will test as much as possible to detect and fix the problems that still exist.
Then I will merge `with_capsul_config` branch to master and this ticket could finally be closed ....  Yippee !!!

In order to work properly, we need some small changes in Capsul, mainly for the module validations. Please merge the [corresponding PR](https://github.com/populse/capsul/pull/220), if it doesn't have any side effects I wouldn't have thought of.

**Comment by denisri** [comment]
I did not get time to test yesterday, I was very busy with other matters. I'm trying now, and I still see some little glitches (on master branch, now):
- remove config.yaml
- run python3 -m populse_mia
- open preferences
- check "Use matlab standalone"
- enter a path for the MCR dir, like `/host/usr/local/spm12-standalone/mcr/v713`
- validate by pressing "OK"
- re-open the preferences
-> the "use matlab standalone" option is not checked any longer. The path is kept but it in disabled.

If I also check "use SPM standalone" and enter a path for it, now the settings are kept.

Now if I check "use matlab", then the SPM standalone option gets disabled, and if I check again "use SPM standalone", then "use matlab" gets disabled and "use matlab standalone" is enabled. Maybe it is the expected behavior, I don't know, but thus it seems impossible to use a "real" matlab for some applications, and SPM standalone at the same time (or the contrary): matlab and spm choices are very tightly linked.

Let's enter a path for matlab executable: check "use matlab" and enter a path for it (SPM standalne gets disabled, as stated just above).

Now I edit the Capsul config, and check the tabs: in "matlab", only the "executable" path is filled, the "mcr_directory" is undefined. Similarly in the SPM config, the directory we have entered for the standalone SPM is not here.

If I enter a path for malab MCR and validate, it does not get into Mia config. It is not possible to "force" the use matlab MCR from Capsul config. And thus it is not possible to have both configs available in capsul for whenever they would be needed. However they actually appear in Mia, thus they do coexist even if disabled.

If, in Capsul config, I go to the SPM tab, check "standalone" and enter paths for spm standalone and matlab MCR, and a SPM version, then validate, now things actually get in Mia. But if I reopen the Capsul config, the SPM version is gone, and this time it's the matlab executable which is not available (same as before, but the other case).

All this is not blocking for me as I don't use matlab at all, I just report a few "strangenesses" I see now.
Maybe the SPM version will be a real issue. I'll check when I have time.

**Comment by servoz** [comment]
No problem @denisri , we do what we can during the days that are unfortunately only 24 hours long!
I am reopening this ticket.

I'll look at what you say as soon as I have some time (I'm very busy at the moment with - among other things - unit tests are currently in fall - calculations in Mia seem to be not working at all ... :-(((().

I merged with master because, _a priori_, there are no more crashes in Mia with this part, the big sync problems seem to be solved now, but I know there may be some left mainly for spm/matlab as I mentioned in a previous post. I wanted to put back the ability to edit the capsul configuration in the mia preferences as soon as possible, for master (I never meant to censor this possibility!)

It does need to be in perfect sync in all cases. 
Thank you for taking the time to do the tests and for reporting the observed problems!

**Comment by servoz** [comment]
Sorry, I've been busy all day with other problems...
I have just tested and reproduced what you have noted!
Having worked a bit on this part, I think I know the corresponding code and it might be more efficient if I look at this.
I'll look into it early next week.

**Comment by servoz** [comment]
I know this part well so I'll take this ticket again (again, again :-))). I'll take a beer when we can finally close it !!!!
A quick look at the codes didn't let me see the error (I was hoping for a big easy to see inconsistency).
So I'll look at it a bit more in detail...

**Comment by servoz** [comment]
I found out where the issue was for the beginning of your report, Denis. I done a commit for this.
I continue to look at the other points you reported in your tests ...

**Comment by servoz** [comment]
> I did not get time to test yesterday, I was very busy with other matters. I'm trying now, and I still see some little glitches (on master branch, now):
    remove config.yaml
    run python3 -m populse_mia
    open preferences
    check "Use matlab standalone"
    enter a path for the MCR dir, like /host/usr/local/spm12-standalone/mcr/v713
    validate by pressing "OK"
    re-open the preferences
    -> the "use matlab standalone" option is not checked any longer. The path is kept but it in disabled.
If I also check "use SPM standalone" and enter a path for it, now the settings are kept.

This now fixed.

> Now if I check "use matlab", then the SPM standalone option gets disabled, and if I check again "use SPM standalone", then "use matlab" gets disabled and "use matlab standalone" is enabled. Maybe it is the expected behavior, I don't know, but thus it seems impossible to use a "real" matlab for some applications, and SPM standalone at the same time (or the contrary): matlab and spm choices are very tightly linked.

That's what we had in mind when we developed this part of the preferences. It has to be simple for the users (this is my main rule for Mia!). That's why there is an internal machinery that automatically changes the buttons. The idea is that if we want to use spm, we must have matlab. If wou want to use spm standalone, you have to use an MCR. So actually, that was a desired feature.

Now there may actually be some trickier cases where the user wants to use pure matlab for some processes and spm standalone (and so MCR) for other processes or other configuration mixes for others. I think this is rarely the case, but that can exist because I have in mind applications that could want this configuration... so I'm not against ...

 What would be really great would be to have a mia config as easy to use and intuitive as possible, a little bit coded as it is now, and access to the capsul config for the more complex cases (as we have started to do with the Edit CAPSUL Config button in Mia pref). And that the two are well synchronized. That's why we really need to discuss the config!

**Comment by servoz** [comment]
> Let's enter a path for matlab executable: check "use matlab" and enter a path for it (SPM standalne gets disabled, as stated just above).
Now I edit the Capsul config, and check the tabs: in "matlab", only the "executable" path is filled, the "mcr_directory" is undefined. Similarly in the SPM config, the directory we have entered for the standalone SPM is not here.
If I enter a path for malab MCR and validate, it does not get into Mia config. It is not possible to "force" the use matlab MCR from Capsul config. And thus it is not possible to have both configs available in capsul for whenever they would be needed. However they actually appear in Mia, thus they do coexist even if disabled.
If, in Capsul config, I go to the SPM tab, check "standalone" and enter paths for spm standalone and matlab MCR, and a SPM version, then validate, now things actually get in Mia. But if I reopen the Capsul config, the SPM version is gone, and this time it's the matlab executable which is not available (same as before, but the other case).

Yes, I understand very well.

It's just that when I made the last changes for the synchronisation I coded things to be more or less the same with capsul config (as the idea for mia pref that i developped just before). I'm not sure that's a good idea after all.

I'd have to remodel the codes to allow all possible configurations in capsul config. In this case the synchronisation will be a bit complicated to code. But why not, maybe it's possible (I admit I haven't coded in this sense).

Or maybe we make a single configuration interface (instead of mia pref + capsul config). In that case, I would like to keep it simple and intuitive.

**Comment by denisri** [comment]
I'm fixing a problem for our morphologist pipeline: I need to fix SPM version variables which are "lost" in duplicated Capsul configs, we need it to correctly use it. It will imply a slight modif in populse_mia (select between existing capsul configs, when some exist). I'm pushing this soon.
But you can close the ticket.

**Comment by servoz** [comment]
After having tested as best as possible, it seems to me that there is only one problem left. The version of matlab is not kept.
For example, if we start from a missing (deleted) configuration file, if we define the matlab and spm executable and the spm version in capsul config (nothing is defined before in mia pref), after coming back to mia pref we find matlab and spm, but if we click again on Edit Capsul config, we lose the version (but everything else is good). This is because there is no version parameter in mia pref for the moment, and as capsul config is recreated from mia pref when we click on the button Edit Caspul config, we lose this information... 
OK @denisri I close this ticket. This version problem will find its place in a new ticket defining the changes to make, including the version management, but in this case the possibility to edit it in the GUI (no hidden conf).

**Comment by denisri** [comment]
I have pushed modifs in capsul and populse_mia which fixes the problem for morphologist. Normally the cpasul settings editor now displays all configs. It's still to be improved but it allows to do the job, at least.

**Comment by servoz** [comment]
OK thanks !

**Comment by servoz** [comment]
Just tested! This is very nice.

Just one thing:
Is it possible to protect by a try the tab.accept() in SettingsEditor.accept() ?

I had protected it to avoid crashing Mia if a bad parameter is entered (typo, etc ...).

For example in caspul config, put an error in the spm path, then do ok -> Mia crash.

Then if we change by:
```
     try:
         tab.accept()
     except Exception as e:
         print(e)

```
by doing the typo, it doesn't crash anymore and a small message appears for the user in the satndard output.

Is it possible to protect again the tab.accept() ?

**Comment by LStruber** [comment]
@servoz synchronization seems to work in both ways for AFNI and ANTS. I also added an AFNI example path in the MIA config window, but not for ANTS since ANTS puts all files in one big folder, that you choose in installation.

**Comment by servoz** [comment]
OK, thank you @LStruber !

**Comment by denisri** [comment]
> Is it possible to protect by a try the tab.accept() in SettingsEditor.accept() ?

yes, sure. I'm doing it.

**Comment by denisri** [comment]
done.

**Comment by servoz** [comment]
Thanks @denisri  !
